File: app/src/main/java/com/amir/stickergram/sticker/pack/user/PackAdapter.java
package com.amir.stickergram.sticker.pack.user;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.amir.stickergram.R;
import com.amir.stickergram.base.BaseActivity;
import com.amir.stickergram.base.BaseFragment;
import com.amir.stickergram.infrastructure.Constants;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class PackAdapter extends RecyclerView.Adapter<ViewHolder> implements View.OnClickListener, View.OnLongClickListener {
    private final OnStickerClickListener listener;
    private final LayoutInflater inflater;
    private List<String> items;
    private String folder;
    private PackItem lastLongClickedItem;
    private String baseThumbDir;
    private String baseDir;

    public PackAdapter(BaseActivity fragment, OnStickerClickListener listener, String folder, String baseDir, String baseThumbDir) {
        this.listener = listener;
        this.folder = folder;
        this.baseDir = baseDir;
        this.inflater = fragment.getLayoutInflater();
        this.baseThumbDir = baseThumbDir;
        items = new ArrayList<>();
        refresh();
    }


    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.item_pack_sticker, parent, false);
        view.setOnClickListener(this);
        view.setOnLongClickListener(this);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.populate(new PackItem(
                folder,
                String.valueOf(position)/*as the name of the file */,
                baseThumbDir,
                baseDir));
    }

    @Override
    public int getItemCount() {
//        Log.e(getClass().getSimpleName(), "size is: " + items.size());
        return items.size();
    }

    @Override
    public void onClick(View view) {
        if (view.getTag() instanceof PackItem) {
            listener.OnIconClicked((PackItem) view.getTag());
        }

    }

    @Override
    public boolean onLongClick(View view) {
        if (view.getTag() instanceof PackItem) {
            lastLongClickedItem = (PackItem) view.getTag();
            listener.OnLongClicked(lastLongClickedItem);
            return true;
        }

        return false;
    }

    void itemRemoved(String dir) {
        lastLongClickedItem.removeThumb();

        int i = items.indexOf(dir);
        notifyItemRemoved(i);
        items.remove(i);
        renameAllFilesAfterThisPosition(i);
        refresh();
        if (items.isEmpty()) {
            File file = new File(baseDir + folder);
            if (file.exists()) {
                file.delete();
                listener.folderDeleted();
            }
        }
    }

    private void renameAllFilesAfterThisPosition(int i) {
        int length = items.size();
        while (i < length) {
            String currentName = baseDir + folder + File.separator + (i + 1) + Constants.PNG;
            File file = new File(currentName);
            if (file.exists()) {
                String newName = baseDir + folder + File.separator + i + Constants.PNG;
                File temp = new File(newName);
                notifyItemChanged(i);
                file.renameTo(temp);
            }
            File thumbFile = new File(baseThumbDir + File.separator + folder + "_" + (i + 1) + Constants.PNG);
            if (thumbFile.exists())
                thumbFile.renameTo(new File(baseThumbDir + File.separator + folder + "_" + i + Constants.PNG));
            i++;
        }
    }

    public void refresh() {
        File folder = new File(baseDir + this.folder + File.separator);
        if (folder.exists()) {
            if (folder.isDirectory()) {
                items = null;
                items = new ArrayList<>();
                File[] files = folder.listFiles();
                for (int i = 0; i < files.length; i++) { //lol ikr you are like why wouldn't list the files in that directory and i'm like cuz listing the file won't return to you a sorted list which is required in order for the delete function work properly
                    items.add(baseDir + this.folder + File.separator + i + Constants.PNG);
                }
            } else
                Log.e(getClass().getSimpleName(), folder.getAbsolutePath() + "was not a directory");
        } else Log.e(getClass().getSimpleName(), folder.getAbsolutePath() + "didn't exist");
    }

    void notifyItemRange() {
        notifyItemRangeChanged(0, items.size());
    }

}

File: app/src/main/java/com/amir/stickergram/sticker/pack/user/PackAdapter.kt
package com.amir.stickergram.sticker.pack.user

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.amir.stickergram.R
import com.amir.stickergram.base.BaseActivity
import com.amir.stickergram.infrastructure.Constants
import java.io.File
import java.util.*

class PackAdapter(fragment: BaseActivity, private val listener: OnStickerClickListener, private val folder: String, private val baseDir: String, private val baseThumbDir: String) : RecyclerView.Adapter<ViewHolder>(), View.OnClickListener, View.OnLongClickListener {
    private val inflater: LayoutInflater = fragment.layoutInflater
    private var items: MutableList<String> = ArrayList()
    private var lastLongClickedItem: PackItem? = null

    init {
        refresh()
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.item_pack_sticker, parent, false)
        view.setOnClickListener(this)
        view.setOnLongClickListener(this)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.populate(PackItem(
                folder,
                position.toString()/*as the name of the file */,
                baseThumbDir,
                baseDir))
    }

    override fun getItemCount(): Int {
        //        Log.e(getClass().getSimpleName(), "size is: " + items.size());
        return items.size
    }

    override fun onClick(view: View) {
        if (view.tag is PackItem) {
            listener.OnIconClicked(view.tag as PackItem)
        }

    }

    override fun onLongClick(view: View): Boolean {
        if (view.tag is PackItem) {
            lastLongClickedItem = view.tag as PackItem
            listener.OnLongClicked(lastLongClickedItem)
            return true
        }

        return false
    }

    fun itemRemoved(dir: String) {
        lastLongClickedItem!!.removeThumb()

        val i = items.indexOf(dir)
        notifyItemRemoved(i)
        items.removeAt(i)
        renameAllFilesAfterThisPosition(i)
        refresh()
        if (items.isEmpty()) {
            val file = File(baseDir + folder)
            if (file.exists()) {
                file.delete()
                listener.folderDeleted()
            }
        }
    }

    private fun renameAllFilesAfterThisPosition(i: Int) {
        var i = i
        val length = items.size
        while (i < length) {
            val currentName = baseDir + folder + File.separator + (i + 1) + Constants.PNG
            val file = File(currentName)
            if (file.exists()) {
                val newName = baseDir + folder + File.separator + i + Constants.PNG
                val temp = File(newName)
                notifyItemChanged(i)
                file.renameTo(temp)
            }
            val thumbFile = File(baseThumbDir + File.separator + folder + "_" + (i + 1) + Constants.PNG)
            if (thumbFile.exists())
                thumbFile.renameTo(File(baseThumbDir + File.separator + folder + "_" + i + Constants.PNG))
            i++
        }
    }

    fun refresh() {
        val folder = File(baseDir + this.folder + File.separator)
        if (folder.exists()) {
            if (folder.isDirectory) {
                items = ArrayList()
                val files = folder.listFiles()
                for (i in files!!.indices) { //lol ikr you are like why wouldn't list the files in that directory and i'm like cuz listing the file won't return to you a sorted list which is required in order for the delete function work properly

                    val pngDirectory = baseDir + this.folder + File.separator + i + Constants.PNG
                    items.add(pngDirectory)

                }
            } else {
                Log.e(javaClass.simpleName, folder.absolutePath + "was not a directory")
            }
        } else {
            Log.e(javaClass.simpleName, folder.absolutePath + "didn't exist")
        }
    }

    fun notifyItemRange() {
        notifyItemRangeChanged(0, items.size)
    }



}

